app/src/main/java/com/lightricks/feedexercise/ui/feed/FeedViewModel.kt

```diff
@@ -1,34 +1,70 @@
 package com.lightricks.feedexercise.ui.feed
 
+import android.util.Log
 import androidx.lifecycle.*
 import com.lightricks.feedexercise.data.FeedItem
 import com.lightricks.feedexercise.data.FeedRepository
 import com.lightricks.feedexercise.network.FeedApiService
 import com.lightricks.feedexercise.util.Event
 import java.lang.IllegalArgumentException
+import com.lightricks.feedexercise.R
 
 /**
  * This view model manages the data for [FeedFragment].
  */
 open class FeedViewModel(private val feedRepository: FeedRepository) : ViewModel() {
-    private val isLoading = MutableLiveData<Boolean>()
-    private val isEmpty = MutableLiveData<Boolean>()
-    private val feedItems = MediatorLiveData<List<FeedItem>>()
+    private val TAG = "FeedViewModel"
+
+    private val stateLiveData = MutableLiveData<State>()
     private val networkErrorEvent = MutableLiveData<Event<String>>()
 
-    fun getIsLoading(): LiveData<Boolean> = isLoading
-    fun getIsEmpty(): LiveData<Boolean> = isEmpty
-    fun getFeedItems(): LiveData<List<FeedItem>> = feedItems
+    fun getIsLoading(): LiveData<Boolean> = Transformations.map(stateLiveData) { it.isLoading }
+    fun getIsEmpty(): LiveData<Boolean> = Transformations.map(stateLiveData) { it.isFeedEmpty() }
+    fun getFeedItems(): LiveData<List<FeedItem>> =
+        Transformations.map(stateLiveData) { it.feedItems }
+
     fun getNetworkErrorEvent(): LiveData<Event<String>> = networkErrorEvent
 
     init {
+        observeRepository()
         refresh()
     }
 
     fun refresh() {
-        //todo: fix the implementation
-        isLoading.value = false
-        isEmpty.value = true
+        setStartLoading()
+        feedRepository.fetchFeed()
+            .subscribe({ setStoppedLoading() },
+                { error -> handleNetworkError(error) })
+    }
+
+    private fun handleNetworkError(error: Throwable?) {
+        networkErrorEvent.postValue(Event(R.string.repository_network_error.toString()))
```

Comment:
In general we should avoid using the postValue method. Replace it by the setValue method. It can cause races in case we use the setValue and the postValue on the same object. So the best practice is to not use it at all. The reason we can do that is that all the method of the view model should run on the main thread and since all the methods run on the main thread the postValue become useless.